src/core/cfg.cpp

```diff
@@ -2274,105 +2222,197 @@ class CFGVisitor : public ASTVisitor {
     }
 
     bool visit_with(AST_With* node) override {
-        char ctxmgrname_buf[80];
-        snprintf(ctxmgrname_buf, 80, "#ctxmgr_%p", node);
-        InternedString ctxmgrname = internString(ctxmgrname_buf);
-        char exitname_buf[80];
-        snprintf(exitname_buf, 80, "#exit_%p", node);
-        InternedString exitname = internString(exitname_buf);
+        // see https://www.python.org/dev/peps/pep-0343/
+        // section "Specification: the 'with' Statement"
+        // which contains pseudocode for what this implements:
+        //
+        // mgr = (EXPR)
+        // exit = type(mgr).__exit__            # not calling it yet
+        // value = type(mgr).__enter__(mgr)
+        // exc = True
+        // try:
+        //     VAR = value
+        //     BLOCK
+        // except:
+        //     exc = False
+        //     if not exit(mgr, *sys.exc_info()):
+        //         raise
+        // finally:
+        //     if exc:
+        //         exit(mgr, None, None, None)
+        //
+        InternedString ctxmgrname = nodeName(node, "ctxmgr");
+        InternedString exitname = nodeName(node, "exit");
+        InternedString whyname = nodeName(node, "why");
+        InternedString exc_type_name = nodeName(node, "exc_type");
+        InternedString exc_value_name = nodeName(node, "exc_value");
+        InternedString exc_traceback_name = nodeName(node, "exc_traceback");
         InternedString nonename = internString("None");
+        CFGBlock* exit_block = cfg->addDeferredBlock();
+        exit_block->info = "with_exit";
 
         pushAssign(ctxmgrname, remapExpr(node->context_expr));
 
-        AST_expr* enter
-            = makeLoadAttribute(makeName(ctxmgrname, AST_TYPE::Load, node->lineno), internString("__enter__"), true);
-        AST_expr* exit
-            = makeLoadAttribute(makeName(ctxmgrname, AST_TYPE::Load, node->lineno), internString("__exit__"), true);
+        // TODO(rntz): for some reason this acts like it's "mgr.__exit__" instead
```

Comment:
So the pseudocode they gave is not quite right in this corner case:
```python
class M(type):
    def __exit__(*args):
        print args
        return False

class C(object):
    __metaclass__ = M

def f(o):
    print hasattr(type(o), "__exit__")
    # type(o).__exit__ exists but this with statement will still throw an "AttributeError: __exit__":
    with o:
        pass

f(C())
```

Even though type(o) has an `__exit__` method, the lookup is actually happening using Python's "special method lookup rules", which is that it looks it up on the class but not the instance *or the metaclass*.  We usually represent this as a "class-only attribute lookup" (what the last argument as `true` means).  I'm not sure there's any way to represent this kind of lookup from Python code, which is why we added a special "ClsAttribute" type of ast node.